backend/business-partner-agent/src/main/java/org/hyperledger/bpa/impl/ActivitiesManager.java

```diff
@@ -0,0 +1,237 @@
+            Boolean task) {
+        List<ActivityItem> results = new ArrayList<>();
+        if (type == null || type == ActivityType.PRESENTATION_EXCHANGE) {
+            Iterable<PartnerProof> proofs = proofRepository.findByStateIn(states);
+            for (PartnerProof p : proofs) {
+                results.add(getPresentationExchangeItem(p, task));
+            }
+        }
+        return results;
+    }
+
+    private ActivityItem getConnectionRequestItem(Partner p, Boolean task) {
+        ActivityRole role = (p.getIncoming() == null) ? ActivityRole.CONNECTION_REQUEST_SENDER
+                : ActivityRole.CONNECTION_REQUEST_RECIPIENT;
+        ActivityType type = ActivityType.CONNECTION_REQUEST;
+
+        ActivityState state;
+        switch (p.getState()) {
+        case ACTIVE:
+        case RESPONSE:
+            state = ActivityState.CONNECTION_REQUEST_ACCEPTED;
+            break;
+        default:
+            switch (role) {
+            case CONNECTION_REQUEST_SENDER:
+                state = ActivityState.CONNECTION_REQUEST_SENT;
+                break;
+            default:
+                state = ActivityState.CONNECTION_REQUEST_RECEIVED;
+            }
+        }
+
+        Long updatedAt = p.getUpdatedAt().toEpochMilli();
+        String linkId = p.getId().toString();
+        return ActivityItem.builder()
+                .role(role)
+                .state(state)
+                .type(type)
+                .partner(partnerManager.refreshPartner(p.getId()).get())
+                .linkId(linkId)
+                .task(task)
+                .updatedAt(updatedAt)
+                .build();
+    }
+
+    private ActivityItem getPresentationExchangeItem(PartnerProof p, Boolean task) {
+        ActivityRole role = p.getRole() == PresentationExchangeRole.PROVER ? ActivityRole.PRESENTATION_EXCHANGE_PROVER
+                : ActivityRole.PRESENTATION_EXCHANGE_VERIFIER;
+        ActivityType type = ActivityType.PRESENTATION_EXCHANGE;
+
+        ActivityState state;
+        switch (p.getState()) {
+        case VERIFIED:
+        case PRESENTATION_ACKED:
+            state = ActivityState.PRESENTATION_EXCHANGE_ACCEPTED;
+            break;
+        case REQUEST_SENT:
+        case PRESENTATIONS_SENT:
+            state = ActivityState.PRESENTATION_EXCHANGE_SENT;
+            break;
+        case REQUEST_RECEIVED:
+        case PRESENTATION_RECEIVED:
+            state = ActivityState.PRESENTATION_EXCHANGE_RECEIVED;
+            break;
+        default:
+            switch (role) {
+            case PRESENTATION_EXCHANGE_PROVER:
+                state = ActivityState.PRESENTATION_EXCHANGE_SENT;
+                break;
+            default:
+                state = ActivityState.PRESENTATION_EXCHANGE_RECEIVED;
+            }
+        }
+        String linkId = p.getId().toString();
+        Optional<PartnerAPI> partner = partnerManager.refreshPartner(p.getPartnerId());
```

Comment:
Partner refresh is a pretty expensive operation which can take a couple of seconds. I do not see a reason for this. A partner proof must always be linked to a partner, also the partner id will not change. If it does something is completely broken.

I would guess this would work as well:

```java
UUID linkId = p.getPartnerId()
PartnerAPI apiPartner;
Optional<Partner> dbPartner = partnerRepo.findById(linkId);
if (dbPartner.isPresent()) {
    apiPartner = converter.toAPIObject(dbPartner); // org.hyperledger.bpa.impl.util.Converter
} 
```

What we need to do at some point is to make the relation Partner to PartnerProof one to many.

Reply:
Yes, refresh takes too long, I only did that because the Frontend logic could return different partner names if a partner is returned without all the data. So if the same partner has ALL it's data (VC, profile etc) then it could get a different name than when it returns with only the basic data.  This is a bigger issue that we should file - I think the partner display name logic should be in the backend and stored in the db. So if the alias hasn't been explicitly set, and we find a profile, then we can set the field to that, and when a Verifiable Presentation is found then use that etc (or whatever all the different ways to calculate the name are).  But for now, I'll speed this up using the above and just returning the basic partner.

Reply:
Then we need to reopen https://github.com/hyperledger-labs/business-partner-agent/issues/77 and agree on a design and name resolution order so that it is consistent everywhere. It's probably best to add a partnerName field to the PartnerAPI and move the name resolution code from the frontend into a strategy class in the backend. But anyway calling refresh while resolving activities is not without side effects.